grafana/resource_annotation.go

```diff
@@ -0,0 +1,174 @@
+package grafana
+
+import (
+	"context"
+	"net/url"
+	"strconv"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceAnnotation() *schema.Resource {
+	return &schema.Resource{
+
+		Description: `
+* [Official documentation](https://grafana.com/docs/grafana/latest/dashboards/annotations/)
+* [HTTP API](https://grafana.com/docs/grafana/latest/http_api/annotations/)
+`,
+
+		CreateContext: CreateAnnotation,
+		UpdateContext: UpdateAnnotation,
+		DeleteContext: DeleteAnnotation,
+		ReadContext:   ReadAnnotation,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"dashboard_id": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     false,
+				Description: "The ID of the dashboard on which to create the annotation.",
+			},
+
+			"panel_id": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     false,
```

Comment:
Should be required
```suggestion
				Required:    true,
```

Reply:
@julienduchesne As mentioned above, the [Grafana API docs](https://grafana.com/docs/grafana/latest/developers/http_api/annotations/#create-annotation) lead me to believe `panel_id` should be optional.

Reply:
Indeed, you're right. Then the default value should just be empty, not `false`. A boolean doesn't make much sense in this context. Also, 
- it probably needs to cause a recreate to change the `panelId` or the `dashboardId`
- I think panelId is required when you set dashboardId and vice-versa. You can use `RequiredWith` for this
- It needs a test case, we need to test that having an annotation without panelId and setting one does recreate the resource correctly

Reply:
> I think panelId is required when you set dashboardId and vice-versa.

...and I do think we need to verify this, as this has not been my understanding, historically.

Reply:
> it probably needs to cause a recreate to change the `panelId` or the `dashboardId`

@julienduchesne Is this to say you'd like to see a `ForceNew: true` associated with the `dashboard_id` and `panel_id` `*schema.Schema`? If so, I've added this.

> I think panelId is required when you set dashboardId and vice-versa.

@julienduchesne I don't believe this is true; I don't believe `panelId` and `dashboardId` are co-dependent, nor are either required. I believe I've proven this via an acceptance test configuration that uses `dashboard_id` without `panel_id`, and an acceptance test configuration that uses `panel_id` without a `dashboard_id`. Do you agree?

> It needs a test case, we need to test that having an annotation without panelId and setting one does recreate the resource correctly

@julienduchesne I believe I've added test cases covering these scenarios. Are there other test cases you'd like to see?